substrate/frame/staking/src/lib.rs

```diff
@@ -1223,3 +1223,67 @@ impl BenchmarkingConfig for TestBenchmarkingConfig {
 	type MaxValidators = frame_support::traits::ConstU32<100>;
 	type MaxNominators = frame_support::traits::ConstU32<100>;
 }
+
+/// Controls validator disabling
+pub trait DisablingStrategy<T: Config> {
+	/// Make a decision if an offender should be disabled or not. The result is a `Vec` of validator
+	/// indices that should be disabled
+	fn make_disabling_decision(
+		offender_stash: &T::AccountId,
+		slash_era: EraIndex,
+		currently_disabled: &Vec<u32>,
+	) -> Option<u32>;
+}
+
+/// Implementation of [`DisablingStrategy`] which disables validators from the active set up to a
+/// threshold. `DISABLING_THRESHOLD_FACTOR` is the factor of the maximum disabled validators in the
+/// active set. E.g. setting this value to `3` means no more than 1/3 of the validators in the
+/// active set can be disabled in an era.
+/// By default a factor of 3 is used which is the byzantine threshold.
+pub struct UpToThresholdDisablingStrategy<const DISABLING_THRESHOLD_FACTOR: usize = 3>;
+
+impl<const DISABLING_THRESHOLD_FACTOR: usize>
+	UpToThresholdDisablingStrategy<DISABLING_THRESHOLD_FACTOR>
+{
+	/// Disabling limit calculated from the total number of validators in the active set. When
+	/// reached no more validators will be disabled.
+	pub fn disable_threshold(validators_len: usize) -> usize {
+		validators_len
+			.saturating_sub(1)
+			.checked_div(DISABLING_THRESHOLD_FACTOR)
+			.unwrap_or_else(|| {
+				defensive!("DISABLING_THRESHOLD_FACTOR should not be 0");
```

Comment:
It would be great to figure out a way to fail at compile time if `DISABLING_THRESHOLD_FACTOR` is 0.

Reply:
Yeah this is a bit sloppy, but I don't have a great alternative. 

One way would be to add an optional an auto-implemented `fn validate(&self) {}` to `DisablingStrategy`, then implement it here which checks being non-zero, and call it in the `fn integrity_test` of `pallet_staking`.

Reply:
What do you think about something like [this](https://github.com/nvzqz/static-assertions/issues/40#issuecomment-1458897730)?

We can emulate a static_assert (assert triggered during compile time as in C++) this way.